### PR TITLE
Update PandasCompat.py to resolve references

### PR DESCRIPTION
**Pull request by raybellwaves**
## Description

This PR allows the PandasCompat sphinx ext to contain resolved references. For example, you can now add intersphinx mapping to the content of the admonition.

### Motivation

I enjoy connecting the PyData communities and this PR allows for more opportunities to use intersphinx mapping to link back to the pandas docs.

### History

I first tried this in a previous PR (https://github.com/rapidsai/cudf/pull/15383#discussion_r1537888240) and commented here (https://github.com/rapidsai/cudf/pull/15383#issuecomment-2028451487) that I may get around to investigating this further. I finally had to time to work on this and made a bit of progress.

### Testing

I created a separate repo for this at https://github.com/raybellwaves/compatsphinxext which deploys straight to https://raybellwaves.github.io/compatsphinxext you can see it's working as expected here: https://raybellwaves.github.io/compatsphinxext/compat.html. You should be able to fork that and tinker pretty quickly.

### Further work

This could be cleaned up (for example I couldn't get the [source] to display in the admonition as I worked from the latest sphinx todo extension (https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/todo.py)). The existing pandas-compat Admonition's could be switched to this if agreed. In addition, the documentation around how to write pandas-compat entries going forward (https://github.com/rapidsai/cudf/blob/branch-24.06/docs/cudf/source/developer_guide/documentation.md#comparing-to-pandas) will also have to be updated.

Longer term the extension could be published and used across RAPIDS libraries where there are differences in compatibility with PyData libraries e.g. pandas, network, scikit-learn to simplify linking to those dos. I'm not sure if I'll have time to work on this though.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: docs/cudf/source/_ext/PandasCompat.py
Comment (bdice) [inline docs/cudf/source/_ext/PandasCompat.py:16]: What does the `_` do here? Typically `_` is used as a placeholder, so I worry that this name will be overwritten or misunderstood by readers. Can we import this as some other name, like `from sphinx.locale import _ as ...` so it's possible to find this?

It looks like it's defined as `_ = get_translation('sphinx')`. Maybe we can use the name `get_translation_sphinx`.

File: docs/cudf/source/_ext/PandasCompat.py
Comment (bdice) [inline docs/cudf/source/_ext/PandasCompat.py:97]: The extra `d` might be a typo?
```suggestion
class PandasCompatDomain(Domain):
```

File: docs/cudf/source/_ext/PandasCompat.py
Comment (bdice) [inline docs/cudf/source/_ext/PandasCompat.py:126]: Do we want to keep the `print` statements in this file, or can they be cleaned up?

File: docs/cudf/source/_ext/PandasCompat.py
Comment (bdice) [inline docs/cudf/source/_ext/PandasCompat.py:132]: Is it possible to write this as a list comprehension? Something like:
```suggestion
        pandascompats = [v for vals in self.domain.pandascompats.values() for v in vals]
```
or alternatively
```python
from itertools import chain
pandascompats = list(chain(*self.domain.pandascompats.values()))
```

File: docs/cudf/source/_ext/PandasCompat.py
Comment (bdice) [inline docs/cudf/source/_ext/PandasCompat.py:133]: Does this need to be materialized into a `list`, or can you use the results of `findall` directly?
```suggestion
        for node in doctree.findall(PandasCompatList):
```

File: docs/cudf/source/_ext/PandasCompat.py
Comment (bdice) [inline docs/cudf/source/_ext/PandasCompat.py:141]:
```suggestion
            content: list[Elements] = [nodes.target()] if node.get("ids") else []
```

**Review by bdice — COMMENTED** [review]
A few initial comments. I don't know the internals of Sphinx that well (though I've dug through them before...), so feel free to let me know if I'm tweaking anything unusual.

File: docs/cudf/source/_ext/PandasCompat.py
Comment (raybellwaves) [reply docs/cudf/source/_ext/PandasCompat.py:16]: Sounds good. `_` is a horrible name for an object

File: docs/cudf/source/_ext/PandasCompat.py
Comment (raybellwaves) [reply docs/cudf/source/_ext/PandasCompat.py:97]: Thanks for the spot

File: docs/cudf/source/_ext/PandasCompat.py
Comment (raybellwaves) [reply docs/cudf/source/_ext/PandasCompat.py:126]: Whoops. These should be removed

File: docs/cudf/source/_ext/PandasCompat.py
Comment (raybellwaves) [reply docs/cudf/source/_ext/PandasCompat.py:132]: I'll take a look

File: docs/cudf/source/_ext/PandasCompat.py
Comment (raybellwaves) [reply docs/cudf/source/_ext/PandasCompat.py:133]: I'll try

**Comment by vyasr** [comment]
@raybellwaves sorry this got a bit lost in the shuffle. @bdice is off for a couple of days so this may not get done until next week, but I've kicked off tests so that we can at least get a doc build going and make sure it works as expected.

**Comment by raybellwaves** [comment]
> @raybellwaves sorry this got a bit lost in the shuffle. @bdice is off for a couple of days so this may not get done until next week, but I've kicked off tests so that we can at least get a doc build going and make sure it works as expected.

Np. Sorry i'm slow to fix as well. I'll ping when ready for another review.

**Comment by raybellwaves** [comment]
Ready for another round of reviews. Thanks for the code simplification suggestions @bdice. It still works as expected as shown here: https://raybellwaves.github.io/compatsphinxext/compat.html

A minor difference is the `[source]` link is now below the admonition (text box). I couldn't work out how to add it into the admonition box. I asked on the sphinx discussion page https://github.com/orgs/sphinx-doc/discussions/12366 but it hasn't been answered. I would say it's minor and only aesthetic and doesn't need to block the PR but of course you can disagree.

**Comment by vyasr** [comment]
@raybellwaves thanks for coming back to this! Given the timing (we're about to freeze the 24.06 release), I'm going to retarget at 24.08 and merge the latest. Hope you don't mind.

**Comment by vyasr** [comment]
@raybellwaves could you fix the style check here? Thanks!

**Review by bdice — COMMENTED** [review]
It looks like docs builds failed in CI. https://github.com/rapidsai/cudf/actions/runs/9291509987/job/25571688171?pr=15704

We'll need to get those fixed before we can merge this. I'm not sure why it failed, but there is a warning:
```
WARNING: Domain 'PandasCompat::pandascompat' has not implemented a `resolve_any_xref` method [myst.domains]
```

**Comment by raybellwaves** [comment]
> It looks like docs builds failed in CI. https://github.com/rapidsai/cudf/actions/runs/9291509987/job/25571688171?pr=15704
> 
> We'll need to get those fixed before we can merge this. I'm not sure why it failed, but there is a warning:
> 
> ```
> WARNING: Domain 'PandasCompat::pandascompat' has not implemented a `resolve_any_xref` method [myst.domains]
> ```

Not obvious why this appeared. I'll got a minimal pipeline i've been testing this on https://github.com/raybellwaves/compatsphinxext/actions/runs/9342573682/job/25710886035 at it runs fine there.

Could be a sphinx version? I have 7.3.7 in my version (https://github.com/raybellwaves/compatsphinxext/actions/runs/9342573682/job/25710886035#step:5:86) whereas the docs env has 6.2.1 (https://github.com/rapidsai/cudf/actions/runs/9291509987/job/25571688171?pr=15704#step:9:483)

**Comment by bdice** [comment]
@raybellwaves I tried to review this but I don't see any changes in the docs. Does this PR affect anything yet, or is it just setup for some future change?

**Comment by bdice** [comment]
The only difference I see is that the `[source]` link was moved out of the box:

## Before

image


## After

image

**Comment by raybellwaves** [comment]
> The only difference I see is that the `[source]` link was moved out of the box:
> 
> ## Before
> image
> 
> ## After
> image

Correct. Follow on PR is https://github.com/rapidsai/cudf/pull/15846. Sorry if I forgot to link it before.

**Comment by bdice** [comment]
Great! That's fine. It looks like the docs built so I'll merge this and we can try it with #15846.

**Review by bdice — APPROVED** [review]

**Pull request merged**
